chain/chain/src/chain.rs

```diff
@@ -1575,7 +1577,7 @@ impl Chain {
                 debug!(target: "chain", "Check orphans: found {} orphans", orphans.len());
                 for orphan in orphans.into_iter() {
                     let block_hash = orphan.hash();
-                    let timer = metrics::BLOCK_PROCESSING_TIME.start_timer();
+                    let success_timer = metrics::BLOCK_PROCESSING_TIME.start_timer();
```

Comment:
maybe we should simply put the timer inside process_block_single?

(or are there too many returns in that function?)

Reply:
A timer outside is simpler, because the function has 5 returns: 4 question marks and the end of the function.

Reply:
It feels odd that we have some metrics measured on the inside, and some on the outside. I'd say yeah, it probably does make sense to split the function as 

```rust
    fn process_block_single<F, F1, F2, F3>(
        &mut self,
        me: &Option<AccountId>,
        block: MaybeValidated<Block>,
        provenance: Provenance,
        mut block_accepted: F,
        mut block_misses_chunks: F1,
        mut orphan_misses_chunks: F2,
        on_challenge: F3,
    ) -> Result<Option<Tip>, Error>
    where
        F: FnMut(AcceptedBlock),
        F1: Copy + FnMut(BlockMissingChunks),
        F2: Copy + FnMut(OrphanMissingChunks),
        F3: FnMut(ChallengeBody),
    {
        metrics::BLOCK_PROCESSED_TOTAL.inc();
        metrics::NUM_ORPHANS.set(self.orphans.len() as i64);
        let res = self.process_block_single_impl(
            me,
            block,
            provenance,
            block_accepted,
            block_misses_chunks,
            orphan_misses_chunks,
            on_challenge,
        );
        if res.is_ok() {
        } else {
        }
        res
    }

    fn process_block_single_impl<F, F1, F2, F3>(
        &mut self,
        me: &Option<AccountId>,
        block: MaybeValidated<Block>,
        provenance: Provenance,
        mut block_accepted: F,
        mut block_misses_chunks: F1,
        mut orphan_misses_chunks: F2,
        on_challenge: F3,
    ) -> Result<Option<Tip>, Error>
    where
        F: FnMut(AcceptedBlock),
        F1: Copy + FnMut(BlockMissingChunks),
        F2: Copy + FnMut(OrphanMissingChunks),
        F3: FnMut(ChallengeBody),
    {

        let prev_head = self.store.head()?;
```

It looks very ugly because the signature is horrendous, but that seems like an orthogonal problem, tried to tackle that a bit in https://github.com/near/nearcore/pull/6206.

Reply:
Done